framework/common/encryption.cpp

```diff
@@ -281,33 +281,59 @@ bool aes_encrypt(const uint8_t *key, const uint8_t *iv, uint8_t *plaintext, int
     return true;
 }
 
-bool aes_decrypt(const uint8_t *key, const uint8_t *iv, uint8_t *data, uint32_t data_len)
+bool aes_decrypt(const uint8_t *key, const uint8_t *iv, uint8_t *ciphertext, int clen,
+                 uint8_t *plaintext, int &plen)
 {
     EVP_CIPHER_CTX *ctx;
+    int len;
 
-    int plen, len;
-    uint8_t buf[128];
+    /*
+     * The decrypt operation will fail if  the final block is not correctly formatted.
+     * This check requires the output buffer to have sufficient room for this check
+     * which is (inl + cipher_block_size), which is clen + 16 in this case (aes128).
+     */
+    if (plen < clen + 16) {
+        LOG(ERROR) << "Insufficient room for final block (padding) format check";
+        return false;
+    }
 
     ctx = EVP_CIPHER_CTX_new();
     if (NULL == ctx) {
+        MAPF_ERR("Failed to create context");
         return false;
     }
+
     if (EVP_DecryptInit_ex(ctx, EVP_aes_128_cbc(), NULL, key, iv) != 1) {
+        MAPF_ERR("EVP_DecryptInit_ex Failed");
+        EVP_CIPHER_CTX_free(ctx);
         return false;
     }
-    EVP_CIPHER_CTX_set_padding(ctx, 0);
 
-    plen = data_len;
-    if (EVP_DecryptUpdate(ctx, data, &plen, data, data_len) != 1 || plen != (int)data_len) {
+    /*
+     * Provide the message to be decrypted, and obtain the decrypted output.
+     * The decrypted data buffer out passed to EVP_DecryptUpdate() should have
+     * sufficient room for (inl + cipher_block_size) bytes.
+     * In this case - plen has to be >= plaintext_len + 16)
+     */
+    if (EVP_DecryptUpdate(ctx, plaintext, &len, ciphertext, clen) != 1) {
+        MAPF_ERR("EVP_DecryptUpdate Failed");
+        EVP_CIPHER_CTX_free(ctx);
         return false;
     }
+    plen = len;
```

Comment:
Note that unlike the encrypt function, this statement is *not* useless, because below we pass len as the argument instead of plen.